File: src/nodes/parenthesized.cpp
#include "nodes/parenthesized.hpp"

namespace nodes {
    NodeParenthesized::NodeParenthesized(Node *expression) {
        this->expression = expression;
    }

    NodeParenthesized::~NodeParenthesized() {
        if (this->expression) delete this->expression;
    }

    const NodeKind NodeParenthesized::GetKind() const {
        return NodeKind::ND_Parenthesized;
    }

    void NodeParenthesized::Show(apic_string indent) {
        printf("%sNodeParenthesized\n", indent.c_str());
        indent.append("  ");

        this->expression->Show(indent);
        indent.pop_back();
        indent.pop_back();
    }

    const nodes::Node *NodeParenthesized::GetExpression() const {
        return this->expression;
    }
}

File: src/nodes/parenthesized.rs
use crate::nodes::node::Node;
use crate::utils::position::Position;

#[derive(Clone)]
pub struct NodeParenthesized {
    position: Position,
    contained: Node,
}

impl NodeParenthesized {
    pub fn init(position: Position, contained: Node) -> NodeParenthesized {
        NodeParenthesized { position, contained }
    }
    
    pub fn get_position(&self) -> &Position {
        &self.position
    }
    
    pub fn show(&self, indent: &mut String, end: char) {
        print!("{indent}NodeParenthesized{end}");

        indent.push(' ');
        indent.push(' ');
        self.contained.show(indent, end);
        indent.pop();
        indent.pop();
    }
    
    pub fn get_contained(&self) -> &Node {
        &self.contained
    }

    pub fn get_mut_contained(&mut self) -> &mut Node {
        &mut self.contained
    }
}